### DISCUSSION TITLE
Constraints for edges

### DISCUSSION THREAD

**Discussion by lvca** [ArcadeData/arcadedb#1026] [Ideas]
Recently asked on Discord, but discussed in many other topics and issues, we'd need a way to create constraints for edges so they can be attached only to specific vertex types.

The most natural way to achieve this is to mimic what OrientDB does: create constraints on `out` and `in` properties. While in OrientDB `out` and `in` are real properties in the document, with ArcadeDB are managed in binary form (much faster), but then exposed to the users as respectively `@out` and `@in` metadata.

To provide constraints on edge we could simply allow the following syntax `alter property Friend.@out linkedtype Account;alter property Friend.@in linkedtype Account` (in OrientDB was `LINKEDCLASS`). This would mean `Friend` edge type can only connect `Account` to another `Account`.

**Comment by lvca** [comment]
This would also allow displaying the graph model in Studio (https://github.com/ArcadeData/arcadedb/discussions/739)

**Reply by gramian** [reply]
I ilke the idea and the syntax, but how about:
* using `only` instead of `linkedtype`?
* allowing arrays here to permit multiple linakable types?

**Reply by malakeel** [reply]
good point. How about inheritance ? Can we allow overriding ? For example:

User --> HasAccount -> Account
SystemUser extends User
Account extends SystemAccount

Do we implicitly allow SystemAccount to link to Account ?

**Reply by lvca** [reply]
Yes, the inheritance will be used, so if you link the base type, all the subtypes will be allowed.

@gramian what about `of type`? Examples (SQL KEYWORD IN UPPER CASE):

```sql
ALTER PROPERTY Friend.@out OF TYPE Account;
CREATE PROPERTY User.data embedded OF TYPE UserData;
CREATE PROPERTY User.keywords list OF TYPE string;
```

So we can use the same for limiting embedded properties, like EMBEDDED and LIST.

WDYT?

**Reply by gramian** [reply]
I can see the appeal of this, but to me the typing of properties feel like constraints, so I thought wrt syntax of `CREATE PROPERTY` like:
```
CREATE PROPERTY User.keywords LIST (type string)
```
hence `only`, but `oftype` would work (for me) too. And the SQL grammar wouldn't have to be touched, right?

**Comment by lvca** [comment]
Moving this discussion to an issue: https://github.com/ArcadeData/arcadedb/issues/1174
